feat: support for CVSSv2

oxisto opened the pull request:
This PR introduces full support for CVSSv2 including temporal and environmental metrics. The style used is somewhat between the v3 and v4 implementation, but leaning more towards v4 where we have a `Vector` struct representing it all and then individual implementations of scores into that vector.

Side node: if interested, I can try to provide the temporal and environmental calculation for v3 as well, in a similar manner.

Fixes #1379

oxisto committed:
cvss: implementation of CVSS 2.0

djc commented:
Honestly not sure I'm motivated to review this code on a voluntary basis. Maybe your organization can sponsor this work?

djc commented on cvss/src/v2/base.rs:
```diff
@@ -0,0 +1,67 @@
+pub use self::{
+    a::AvailabilityImpact, ac::AccessComplexity, au::Authentication, av::AccessVector,
+    c::ConfidentialityImpact, i::IntegrityImpact,
+};
```
Instead of these `pub use self::` imports, format like this:

```rust
mod a;
pub use a::AvailabilityImpact;

mod ac;
pub use ac::AccessComplexity;
```

etc.

oxisto replied:
Fine by me, I was actually just following the style in the existing 3.0 files. I can probably adjust this in 3.0 in the other PR as well.

djc commented on cvss/src/v2/base/a.rs:
```diff
@@ -0,0 +1,77 @@
+use crate::Error;
+use crate::v2::{Metric, MetricType};
+use alloc::borrow::ToOwned;
+use core::{fmt, str::FromStr};
```
Keep imports in three blocks separated by empty lines.

- `core`, `alloc` and `std`
- External crates
- `crate`

In that order.

djc commented on cvss/src/v2/metric/base/a.rs:
Nit: move associated `const`s to the bottom of the impl block(s).

oxisto replied:
Are you sure about that? That would explicitly not match the structure how it is in v3 and v4?

djc replied:
Yes, I'm sure -- fine with the discrepancy for now.

djc commented on cvss/tests/redhat-v2.rs:
```diff
@@ -0,0 +1,38 @@
+#[test]
+fn cvss_v2_random() {
+    run_tests_from_file("vectors_random2");
+}
```
Nit: tests before the helper function, please.

djc reviewed (commented):
Here's some initial feedback. I'd like this new code to conform to a common style that might not fully align with what's there currently in the crate.

You'll want to rebase on top of `main` and fix any lints (in particular, `clippy::use_self`). `cargo clippy --fix` should be able to take care of most of that.

djc commented on cvss/src/v2/base.rs:
```diff
@@ -0,0 +1,67 @@
+use super::Score;
+use crate::v2::{Metric, Vector};
+
+impl Vector {
```
Let's move all the `Vector` methods into a single `impl` block in `vector.rs`, directly below the type definition.

djc commented on cvss/src/v2/base.rs:
```diff
@@ -0,0 +1,67 @@
+    ///
+    /// Described in CVSS v2.0 Specification: Section 3.2.1:
+    /// <https://www.first.org/cvss/v2/guide#3-2-1-Base-Equation>
+    #[cfg(feature = "std")]
```
Why do we need `std` for these?

oxisto replied:
Because of `std::f64::round`

djc commented:
Please squash all the commits into a single one.

djc commented:
(Add a separate commit with a version bump to fix the CI semver job.)

oxisto commented:
> Please squash all the commits into a single one.

I addressed all the code review comments and did some additional internal audit, that uncovered a few missing pieces, such as integration into OSV. I have kept those changes on a separate commit yet, might make it easier to review.

I did uncover the difference in scoring between FIRST and NIST. FIRST specifies that all the sub scores such  `BaseScore` and `TemperalScore` are rounded to 1 and uses the rounded version to keep calculating (e.g. the temporal and environmental score). For some reason, the NVD NIST calculator takes the unrounded sub score (e.g. from temporal) in the environmental score and then only applies rounding at the end. This produces slightly different results in environmental and temporal scores.

This library currently implements it according to the FIRST spec.

(cc @tschmidtb51)

oxisto committed:
cvss: bumped version to 3.0.0

djc commented on cvss/src/v2/metric.rs:
```diff
@@ -0,0 +1,164 @@
+//! CVSS v2.0 metrics.
+
+use crate::{Error, Result};
```
Nit: keep the `crate` imports in a separate block.

djc commented on cvss/src/v2/metric.rs:
```diff
@@ -0,0 +1,164 @@
+/// Trait for CVSS v2.0 metrics.
+pub trait Metric: Copy + Clone + Debug + Display + Eq + FromStr + Ord {
+    /// [`MetricType`] of this metric.
+    const TYPE: MetricType;
```
Nit: keep the `const` at the bottom of the block (including in all impls).

djc commented on cvss/src/v2/vector.rs:
```diff
@@ -0,0 +1,499 @@
+use {
+    alloc::string::{String, ToString},
+    serde::{Deserialize, Serialize, de, ser},
+};
```
Nit: keep one line per top-level crate. The `alloc` imports should go in the core/alloc/std block, above. (Just duplicate the guard.)

djc commented on cvss/src/v2/vector.rs:
Can we just write `Self::default()` here?

djc commented on cvss/src/v2/vector.rs:
Nit: write `T::from_str(new_val)?`, please, without the type annotation.

djc commented on cvss/Cargo.toml:
```diff
@@ -1,20 +1,21 @@
 readme       = "README.md"
 categories   = ["parser-implementations"]
-keywords     = ["cvssv3", "cvssv4", "security", "advisory", "vulnerability"]
+keywords     = ["cvssv2","cvssv3", "cvssv4", "security", "advisory", "vulnerability"]
```
Nit: add a space after the new keyword.

djc reviewed (approved):
This mostly looks good to me. Please squash all the commits except the version bump, and address the remaining nits.

djc commented:
Thanks!
